llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -13221,3 +13237,17 @@ bool RISCVTargetLowering::getPostIndexedAddressParts(SDNode *N, SDNode *Op,
   AM = ISD::POST_INC;
   return true;
 }
+
+void RISCVTargetLowering::getAddressParts(SelectionDAG &DAG, const SDLoc &DL, SDValue Addr,
+                                          SDValue &Base, SDValue &Offset, EVT PtrVT) const {
+  if (auto *G = dyn_cast<GlobalAddressSDNode>(Addr)) {
+    Base = DAG.getTargetGlobalAddress(G->getGlobal(), DL, PtrVT);
+    Offset = DAG.getConstant(G->getOffset(), DL, PtrVT);
+  } else if (auto *C = dyn_cast<ConstantSDNode>(Addr)) {
+    Base = DAG.getConstant(0, DL, PtrVT);
+    Offset = DAG.getConstant(C->getZExtValue(), DL, PtrVT);
```

Comment:
If the constant value here is greater than 12 bits, we might end up being unable to represent it in the immediate field of the `elw` instruction. Perhaps a suggestion to fix this would be to split up the constant so that only the lower 12 bits get reported as `Offset`, and the higher bits are loaded into `Base`. There may be better ways to do this that match how we already handle this for RISC-V loads